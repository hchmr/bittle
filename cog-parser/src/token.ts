import { Point } from './tree.js';

export type Token = {
    kind: TokenKind;
    lexeme: string;
    startPosition: Point;
    leadingTrivia: string[];
    trailingTrivia: string[];
};

export type TokenKind =
    | '<eof>'
    | '<identifier>'
    | '<int>'
    | '<string>'
    | '<char>'
    | '<error>'
    | (typeof symbols)[number]
    | (typeof keywords)[number];

export const symbols = [
    '(',
    ')',
    '{',
    '}',
    '[',
    ']',
    ':',
    ';',
    '.',
    ',',
    '...',
    '~',
    '|',
    '||',
    '&',
    '&&',
    '>',
    '>>',
    '>=',
    '>>=',
    '=',
    '==',
    '=>',
    '!',
    '!=',
    '?',
    '<',
    '<<',
    '<=',
    '<-',
    '<<=',
    '+',
    '+=',
    '-',
    '-=',
    '->',
    '*',
    '/',
    '%',
    '^',
] as const;

export const keywords = [
    'as',
    'break',
    'const',
    'continue',
    'do',
    'else',
    'enum',
    'extern',
    'func',
    'if',
    'include',
    'return',
    'sizeof',
    'struct',
    'var',
    'while',
] as const;
