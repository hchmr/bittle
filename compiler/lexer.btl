include "bittle.btls";

//==============================================================================
//== Helper functions

func try_keyword(lexeme: *Char, keyword: *Char, out: *TokKind, kind: TokKind): Bool {
    if (str_eq(lexeme, keyword)) {
        *out = kind;
        return true;
    } else {
        return false;
    }
}

func lookup_keyword(lexeme: *Char, out: *TokKind): Bool {
    return try_keyword(lexeme, "as", out, Tok_As)
        || try_keyword(lexeme, "break", out, Tok_Break)
        || try_keyword(lexeme, "const", out, Tok_Const)
        || try_keyword(lexeme, "continue", out, Tok_Continue)
        || try_keyword(lexeme, "else", out, Tok_Else)
        || try_keyword(lexeme, "enum", out, Tok_Enum)
        || try_keyword(lexeme, "extern", out, Tok_Extern)
        || try_keyword(lexeme, "false", out, Tok_False)
        || try_keyword(lexeme, "for", out, Tok_For)
        || try_keyword(lexeme, "func", out, Tok_Func)
        || try_keyword(lexeme, "if", out, Tok_If)
        || try_keyword(lexeme, "include", out, Tok_Include)
        || try_keyword(lexeme, "null", out, Tok_Null)
        || try_keyword(lexeme, "return", out, Tok_Return)
        || try_keyword(lexeme, "sizeof", out, Tok_Sizeof)
        || try_keyword(lexeme, "struct", out, Tok_Struct)
        || try_keyword(lexeme, "true", out, Tok_True)
        || try_keyword(lexeme, "var", out, Tok_Var)
        || try_keyword(lexeme, "while", out, Tok_While);
}

//==============================================================================
//== Lexer core

struct Lexer {
    input: *Char,
    index: Int,
    pos: Pos,
    tok_kind: TokKind,
    tok_index: Int,
    tok_pos: Pos,
    tok_lexeme: *Char,
}

func lexer_new(file_name: *Char, input: *Char): *Lexer {
    var self: *Lexer = calloc(1, sizeof(Lexer));

    var pos = Pos { file: file_name, row: 1, col: 1 };

    *self = Lexer {
        input: input,
        index: 0,
        pos: pos,
        tok_kind: Tok_Eof,
        tok_index: 0,
        tok_pos: pos,
        tok_lexeme: "",
    };
    return self;
}

func make_lexeme(self: *Lexer): *Char {
    var len = self.index - self.tok_index;
    var lexeme: *Char = malloc(len + 1);
    memcpy(lexeme, &self.input[self.index - len], len);
    lexeme[len] = '\0';
    return lexeme;
}

func begin_token(self: *Lexer) {
    self.tok_index = self.index;
    self.tok_pos = self.pos;
}

func finish_token(self: *Lexer, tok_kind: TokKind, lexeme: *Char) {
    self.tok_kind = tok_kind;
    self.tok_lexeme = lexeme;
}

func curr_char(self: *Lexer): Int32 {
    return self.input[self.index];
}

func next_char(self: *Lexer) {
    var c = curr_char(self);
    if (c == '\n') {
        self.pos.row += 1;
        self.pos.col = 1;
    } else {
        self.pos.col += 1;
    }
    self.index += 1;
}

//==============================================================================
//== Character matching

func at_end(self: *Lexer): Bool {
    return curr_char(self) == '\0';
}

func at_whitespace(self: *Lexer): Bool {
    return is_whitespace(curr_char(self));
}

func at_word_start(self: *Lexer): Bool {
    return is_word_start(curr_char(self));
}

func at_word_part(self: *Lexer): Bool {
    return is_word_part(curr_char(self));
}

func at_digit(self: *Lexer, base: Int): Bool {
    return is_digit(curr_char(self), base);
}

func at_printable(self: *Lexer): Bool {
    return is_print(curr_char(self));
}

func at_char(self: *Lexer, c: Int32): Bool {
    return curr_char(self) == c;
}

func eat_char(self: *Lexer, c: Int32): Bool {
    if (at_char(self, c)) {
        next_char(self);
        return true;
    } else {
        return false;
    }
}

//==============================================================================
//== Scanning routines

func skip_whitespace(self: *Lexer) {
    while (is_whitespace(curr_char(self))) {
        next_char(self);
    }
}

func skip_line_comment(self: *Lexer) {
    while (!at_end(self) && !eat_char(self, '\n')) {
        next_char(self);
    }
}

func skip_block_comment(self: *Lexer) {
    while (true) {
        if (at_end(self)) {
            die_at(&self.pos, "Unterminated block comment.");
        } else if (eat_char(self, '*') && eat_char(self, '/')) {
            break;
        } else {
            next_char(self);
        }
    }
}

func scan_word(self: *Lexer) {
    next_char(self);
    while (is_word_part(curr_char(self))) {
        next_char(self);
    }
    var lexeme = make_lexeme(self);
    var tok_kind = Tok_Ident;
    lookup_keyword(lexeme, &tok_kind);
    finish_token(self, tok_kind, lexeme);
}

func scan_number(self: *Lexer) {
    var base = 10;
    if (eat_char(self, '0')) {
        if (eat_char(self, 'x')) {
            base = 16;
        } else if (eat_char(self, 'b')) {
            base = 2;
        } else if (eat_char(self, 'o')) {
            base = 8;
        }
    }

    if (base != 10 && !at_digit(self, base)) {
        die_at(&self.pos, "Expected base-%d digit.\n", base);
    }

    while (at_digit(self, base)) {
        next_char(self);
    }

    var lexeme = make_lexeme(self);
    finish_token(self, Tok_Int, lexeme);
}

func scan_char_part(self: *Lexer) {
    if (eat_char(self, '\\')) {
        if (eat_char(self, '\'')) {
        } else if (eat_char(self, '\"')) {
        } else if (eat_char(self, '0')) {
        } else if (eat_char(self, '\\')) {
        } else if (eat_char(self, 'n')) {
        } else if (eat_char(self, 'r')) {
        } else if (eat_char(self, 't')) {
        } else {
            die_at(&self.pos, "Invalid escape sequence.");
        }
    } else if (at_printable(self)) {
        next_char(self);
    } else {
        die_at(&self.pos, "Illegal character.\n");
    }
}

func scan_string(self: *Lexer) {
    next_char(self);
    while (!eat_char(self, '\"')) {
        if (at_end(self)) {
            die_at(&self.pos, "Unterminated string literal.");
        } else {
            scan_char_part(self);
        }
    }
    var lexeme = make_lexeme(self);
    finish_token(self, Tok_String, lexeme);
}

func scan_char(self: *Lexer) {
    next_char(self);
    if (at_char(self, '\'')) {
        die_at(&self.pos, "Empty character literal.");
    }
    scan_char_part(self);
    if (!eat_char(self, '\'')) {
        die_at(&self.pos, "Unterminated character literal.");
    }
    var lexeme = make_lexeme(self);
    finish_token(self, Tok_Char, lexeme);
}

func lexer_next(self: *Lexer): Tok {
    while (true) {
        begin_token(self);
        if (at_end(self)) {
            finish_token(self, Tok_Eof, "");
        } else if (at_whitespace(self)) {
            skip_whitespace(self);
            continue;
        } else if (at_word_start(self)) {
            scan_word(self);
        } else if (at_digit(self, 10)) {
            scan_number(self);
        } else if (at_char(self, '\"')) {
            scan_string(self);
        } else if (at_char(self, '\'')) {
            scan_char(self);
        } else if (eat_char(self, '(')) {
            finish_token(self, Tok_LParen, "(");
        } else if (eat_char(self, ')')) {
            finish_token(self, Tok_RParen, ")");
        } else if (eat_char(self, '[')) {
            finish_token(self, Tok_LBracket, "[");
        } else if (eat_char(self, ']')) {
            finish_token(self, Tok_RBracket, "]");
        } else if (eat_char(self, '{')) {
            finish_token(self, Tok_LBrace, "{");
        } else if (eat_char(self, '}')) {
            finish_token(self, Tok_RBrace, "}");
        } else if (eat_char(self, '+')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_PlusEq, "+=");
            } else {
                finish_token(self, Tok_Plus, "+");
            }
        } else if (eat_char(self, '-')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_MinusEq, "-=");
            } else if (eat_char(self, '>')) {
                finish_token(self, Tok_Arrow, "->");
            } else {
                finish_token(self, Tok_Minus, "-");
            }
        } else if (eat_char(self, '*')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_StarEq, "*=");
            } else {
                finish_token(self, Tok_Star, "*");
            }
        } else if (eat_char(self, '/')) {
            if (eat_char(self, '/')) {
                skip_line_comment(self);
                continue;
            } else if (eat_char(self, '*')) {
                skip_block_comment(self);
                continue;
            } else if (eat_char(self, '=')) {
                finish_token(self, Tok_SlashEq, "/=");
            } else {
                finish_token(self, Tok_Slash, "/");
            }
        } else if (eat_char(self, '%')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_PercentEq, "%=");
            } else {
                finish_token(self, Tok_Percent, "%");
            }
        } else if (eat_char(self, '=')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_EqEq, "==");
            } else {
                finish_token(self, Tok_Eq, "=");
            }
        } else if (eat_char(self, '!')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_BangEq, "!=");
            } else {
                finish_token(self, Tok_Bang, "!");
            }
        } else if (eat_char(self, '<')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_LtEq, "<=");
            } else if (eat_char(self, '<')) {
                if (eat_char(self, '=')) {
                    finish_token(self, Tok_LtLtEq, "<<=");
                } else {
                    finish_token(self, Tok_LtLt, "<<");
                }
            } else {
                finish_token(self, Tok_Lt, "<");
            }
        } else if (eat_char(self, '>')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_GtEq, ">=");
            } else if (eat_char(self, '>')) {
                if (eat_char(self, '=')) {
                    finish_token(self, Tok_GtGtEq, ">>=");
                } else {
                    finish_token(self, Tok_GtGt, ">>");
                }
            } else {
                finish_token(self, Tok_Gt, ">");
            }
        } else if (eat_char(self, '&')) {
            if (eat_char(self, '&')) {
                finish_token(self, Tok_AmpAmp, "&&");
            } else if (eat_char(self, '=')) {
                finish_token(self, Tok_AmpEq, "&=");
            } else {
                finish_token(self, Tok_Amp, "&");
            }
        } else if (eat_char(self, '|')) {
            if (eat_char(self, '|')) {
                finish_token(self, Tok_BarBar, "||");
            } else if (eat_char(self, '=')) {
                finish_token(self, Tok_BarEq, "|=");
            } else {
                finish_token(self, Tok_Bar, "|");
            }
        } else if (eat_char(self, '^')) {
            if (eat_char(self, '=')) {
                finish_token(self, Tok_CaretEq, "^=");
            } else {
                finish_token(self, Tok_Caret, "^");
            }
        } else if (eat_char(self, '~')) {
            finish_token(self, Tok_Tilde, "~");
        } else if (eat_char(self, '?')) {
            finish_token(self, Tok_Question, "?");
        } else if (eat_char(self, ',')) {
            finish_token(self, Tok_Comma, ",");
        } else if (eat_char(self, ';')) {
            finish_token(self, Tok_Semicolon, ";");
        } else if (eat_char(self, ':')) {
            if (eat_char(self, ':')) {
                finish_token(self, Tok_ColonColon, "::");
            } else {
                finish_token(self, Tok_Colon, ":");
            }
        } else if (eat_char(self, '.')) {
            if (eat_char(self, '.')) {
                if (eat_char(self, '.')) {
                    finish_token(self, Tok_DotDotDot, "...");
                } else {
                    finish_token(self, Tok_DotDot, "..");
                }
            } else {
                finish_token(self, Tok_Dot, ".");
            }
        } else {
            die_at(&self.pos, "Unexpected character: '%c'.\n", curr_char(self));
        }

        return Tok {
            kind: self.tok_kind,
            pos: self.tok_pos,
            lexeme: self.tok_lexeme,
        };
    }
}
