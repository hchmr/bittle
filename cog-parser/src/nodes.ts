
//=========================================================================
//== Syntax nodes

import { Position, Token, TokenKind } from "./token.js";

export type SyntaxNode =
    | LeafNode
    | CompositeNode;

export type LeafNode =
    | TokenNode
    | MissingTokenNode;

export type CompositeNode = {
    nodeKind: CompositeNodeKind;
    children: SyntaxNodeChild[];
    startPosition: Position;
};

export type SyntaxNodeChild = {
    field?: string;
    node: SyntaxNode;
};

export type TokenNode = {
    token: Token;
};

export type MissingTokenNode = {
    missing: TokenKind;
    position: Position;
}

export enum CompositeNodeKind {
    // Root
    Root = 'Root',
    // Top-level declarations
    IncludeDecl = 'IncludeDecl',
    EnumDecl = 'EnumDecl',
    EnumMember = 'EnumMember',
    StructDecl = 'StructDecl',
    StructMember = 'StructMember',
    FuncDecl = 'FuncDecl',
    FuncParam = 'FuncParam',
    GlobalDecl = 'GlobalDecl',
    ConstDecl = 'ConstDecl',
    // Statements
    BlockStmt = 'BlockStmt',
    VarStmt = 'VarStmt',
    IfStmt = 'IfStmt',
    WhileStmt = 'WhileStmt',
    ReturnStmt = 'ReturnStmt',
    BreakStmt = 'BreakStmt',
    ContinueStmt = 'ContinueStmt',
    ExprStmt = 'ExprStmt',
    // Expressions
    GroupExpr = 'GroupExpr',
    NameExpr = 'NameExpr',
    LiteralExpr = 'LiteralExpr',
    UnaryExpr = 'UnaryExpr',
    BinaryExpr = 'BinaryExpr',
    TernaryExpr = 'TernaryExpr',
    CallExpr = 'CallExpr',
    IndexExpr = 'IndexExpr',
    FieldExpr = 'FieldExpr',
    CastExpr = 'CastExpr',
    SizeofExpr = 'SizeofExpr',
    // Types
    GroupType = 'GroupType',
    NameType = 'NameType',
    PointerType = 'PointerType',
    ArrayType = 'ArrayType',
    // Literals
    IntLiteral = 'IntLiteral',
    StringLiteral = 'StringLiteral',
    CharLiteral = 'CharLiteral',
    // Error
    Error = 'Error',
}

//=========================================================================
//== Utility functions

export function isCompositeNode(node: SyntaxNode): node is CompositeNode {
    if ('nodeKind' in node) {
        assertType<CompositeNodeKind>(node.nodeKind);
        return true;
    } else {
        return false;
    }
}

export function isTokenNode(node: SyntaxNode): node is TokenNode {
    if ('token' in node) {
        assertType<TokenNode>(node);
        return true;
    } else {
        return false;
    }
}

export function isMissingTokenNode(node: SyntaxNode): node is MissingTokenNode {
    if ('missing' in node) {
        assertType<MissingTokenNode>(node);
        return true;
    } else {
        return false;
    }
}

function assertType<T>(_: T): void {
}

export function prettySyntaxTree(node: SyntaxNode): string {
    let text = '';
    (function go(node: SyntaxNode, level: number = 0) {
        if (isCompositeNode(node)) {
            text += indent(level);
            text += node.nodeKind;
            text += prettyPosition(node.startPosition);
            text += '\n';
            for (const child of node.children) {
                go(child.node, level + 1);
            }
        } else if (isTokenNode(node)) {
            text += indent(level);
            text += JSON.stringify(node.token.lexeme);
            text += prettyPosition(node.token.position);
            text += '\n';
        } else if (isMissingTokenNode(node)) {
            text += indent(level);
            text += 'MISSING ' + JSON.stringify(node.missing);
            text += prettyPosition(node.position);
            text += '\n';
        } else {
            const never: never = node;
            throw new Error(`Unreachable`, never);
        }
    })(node);
    return text;

    function prettyPosition(pos: Position) {
        return `@${pos.row + 1}:${pos.col + 1}`;
    }

    function indent(level: number) {
        return '  '.repeat(level);
    }
}

export function reconstructText(node: SyntaxNode): string {
    if (isCompositeNode(node)) {
        return node.children
            .map(child => reconstructText(child.node))
            .join('');
    } else if (isTokenNode(node)) {
        let text = '';
        text += node.token.leadingTrivia.join('');
        text += node.token.lexeme;
        text += node.token.trailingTrivia.join('');
        return text;
    } else if (isMissingTokenNode(node)) {
        return '';
    } else {
        const never: never = node;
        throw new Error(`Unreachable`, never);
    }
}
