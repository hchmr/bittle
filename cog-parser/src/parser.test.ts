import { expect, test } from 'vitest'
import { ErrorSink, Parser } from './parser.js';
import { tokenize } from './lexer.js';
import { reconstructText, prettySyntaxTree, SyntaxNode } from './nodes.js';
import { Position } from './token.js';

type SyntaxError = { pos: Position, message: string };

function parse(text: string): [SyntaxNode, SyntaxError[]] {
    const errors: SyntaxError[] = [];


    const parser = new Parser(
        tokenize(text),
        <ErrorSink>{
            add(pos, message: string) {
                errors.push({ pos, message });
            }
        }
    );

    const tree = parser.top();

    return [tree, errors];
}

const good = `
struct Point {
    x: Int,
    y: Int,
}
func sqr_dst(p1: *Point, p2: *Point): Int {
    var dx = p2.x - p1.x;
    var dy = p2.y - p1.y;
    return dx * dx + dy * dy;
}
`

const bad = `
func add(a: Int, b: Int): Int {

func maine( {
    print("Helo wolrdn\\n");
`

test('parser:good', () => {
    const [tree, errors] = parse(good);
    expect(errors).toStrictEqual([]);
});

test('parser:bad', () => {
    const [tree, errors] = parse(bad);
    expect(errors).toMatchSnapshot();
});

test('reconstructText:good', () => {
    const [tree, _errors] = parse(good);
    expect(reconstructText(tree)).toEqual(good);
})

test('reconstructText:bad', () => {
    const [tree, _errors] = parse(bad);
    expect(reconstructText(tree)).toEqual(bad);
});

test('prettySyntaxTree:good', () => {
    const [tree, _errors] = parse(good);
    expect(prettySyntaxTree(tree)).toMatchSnapshot();
})

test('prettySyntaxTree:bad', () => {
    const [tree, _errors] = parse(bad);
    expect(prettySyntaxTree(tree)).toMatchSnapshot();
})
