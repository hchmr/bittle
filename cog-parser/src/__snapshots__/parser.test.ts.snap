// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parser:bad 1`] = `
[
  {
    "message": "Unexpected start of statement: 'func'.",
    "position": {
      "column": 0,
      "index": 34,
      "row": 3,
    },
  },
  {
    "message": "Expected parameter.",
    "position": {
      "column": 12,
      "index": 46,
      "row": 3,
    },
  },
  {
    "message": "Unexpected start of statement: '<eof>'.",
    "position": {
      "column": 27,
      "index": 75,
      "row": 4,
    },
  },
]
`;

exports[`prettySyntaxTree:bad 1`] = `
"Root@2:1
  FuncDecl@2:1
    "func"@2:1
    "add"@2:6
    "("@2:9
    FuncParam@2:10
      "a"@2:10
      ":"@2:11
      NameType@2:13
        "Int"@2:13
    ","@2:16
    FuncParam@2:18
      "b"@2:18
      ":"@2:19
      NameType@2:21
        "Int"@2:21
    ")"@2:24
    ":"@2:25
    NameType@2:27
      "Int"@2:27
    BlockStmt@2:31
      "{"@2:31
      MISSING "}"@4:1
  FuncDecl@4:1
    "func"@4:1
    "maine"@4:6
    "("@4:11
    MISSING ","@4:13
    MISSING ")"@4:13
    BlockStmt@4:13
      "{"@4:13
      ExprStmt@5:5
        CallExpr@5:5
          NameExpr@5:5
            "print"@5:5
          "("@5:10
          StringLiteral@5:11
            "\\"Helo wolrdn\\\\n\\""@5:11
          ")"@5:26
        ";"@5:27
      Error@5:28
        ""@5:28
      MISSING "}"@5:28
  ""@5:28
"
`;

exports[`prettySyntaxTree:good 1`] = `
"Root@2:1
  StructDecl@2:1
    "struct"@2:1
    "Point"@2:8
    "{"@2:14
    StructMember@3:5
      "x"@3:5
      ":"@3:6
      NameType@3:8
        "Int"@3:8
    ","@3:11
    StructMember@4:5
      "y"@4:5
      ":"@4:6
      NameType@4:8
        "Int"@4:8
    ","@4:11
    "}"@5:1
  FuncDecl@6:1
    "func"@6:1
    "sqr_dst"@6:6
    "("@6:13
    FuncParam@6:14
      "p1"@6:14
      ":"@6:16
      PointerType@6:18
        "*"@6:18
        NameType@6:19
          "Point"@6:19
    ","@6:24
    FuncParam@6:26
      "p2"@6:26
      ":"@6:28
      PointerType@6:30
        "*"@6:30
        NameType@6:31
          "Point"@6:31
    ")"@6:36
    ":"@6:37
    NameType@6:39
      "Int"@6:39
    BlockStmt@6:43
      "{"@6:43
      VarStmt@7:5
        "var"@7:5
        "dx"@7:9
        "="@7:12
        BinaryExpr@7:14
          FieldExpr@7:14
            NameExpr@7:14
              "p2"@7:14
            "."@7:16
            "x"@7:17
          "-"@7:19
          FieldExpr@7:21
            NameExpr@7:21
              "p1"@7:21
            "."@7:23
            "x"@7:24
        ";"@7:25
      VarStmt@8:5
        "var"@8:5
        "dy"@8:9
        "="@8:12
        BinaryExpr@8:14
          FieldExpr@8:14
            NameExpr@8:14
              "p2"@8:14
            "."@8:16
            "y"@8:17
          "-"@8:19
          FieldExpr@8:21
            NameExpr@8:21
              "p1"@8:21
            "."@8:23
            "y"@8:24
        ";"@8:25
      ReturnStmt@9:5
        "return"@9:5
        BinaryExpr@9:12
          BinaryExpr@9:12
            NameExpr@9:12
              "dx"@9:12
            "*"@9:15
            NameExpr@9:17
              "dx"@9:17
          "+"@9:20
          BinaryExpr@9:22
            NameExpr@9:22
              "dy"@9:22
            "*"@9:25
            NameExpr@9:27
              "dy"@9:27
        ";"@9:29
      "}"@10:1
  ""@10:2
"
`;
